library/nist_kw.c

```diff
@@ -0,0 +1,710 @@
+    *out_len = 0;
+
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the ciphertext length for KW
+         * must be between 3 to 2^54 semiblocks inclusive.
+         */
+        if( in_len < 24
+#if SIZE_MAX > 0x200000000000000
+            || in_len > 0x200000000000000
+#endif
+            )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH,
+                      output, out_len );
+        if( ret != 0 )
+            goto cleanup;
+
+        if ( memcmp( ICV1, output, KW_SEMIBLOCK_LENGTH ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        /*
+         * shift output to point to P
+         */
+        for( i = 0; i < ( *out_len / KW_SEMIBLOCK_LENGTH ); i++ )
+        {
+            memcpy( output + ( i * KW_SEMIBLOCK_LENGTH ),
+                    output + ( ( i + 1 ) * KW_SEMIBLOCK_LENGTH ),
+                    KW_SEMIBLOCK_LENGTH );
+        }
+        *out_len = *out_len - KW_SEMIBLOCK_LENGTH;
+    }
+    else if( mode == MBEDTLS_KW_MODE_KWP )
+    {
+        int padlen = 0;
+        uint32_t Plen;
+        /*
+         * According to SP800-38F Table 1, the ciphertext length for KWP
+         * must be between 2 to 2^29 semiblocks inclusive.
+         */
+        if( in_len < KW_SEMIBLOCK_LENGTH * 2
+#if SIZE_MAX > 0x100000000
+            || in_len > 0x100000000
+#endif
+          )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        if( in_len == KW_SEMIBLOCK_LENGTH * 2 )
+        {
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                         input, 16, output, out_len );
+            if( ret != 0 )
+                goto cleanup;
+        }
+        else if( in_len >  KW_SEMIBLOCK_LENGTH * 2 )
+        {
+            ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH,
+                          output, out_len );
+            if( ret != 0 )
+                goto cleanup;
+        }
+        if( memcmp( ICV2, output, KW_SEMIBLOCK_LENGTH / 2 ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        GET_UINT32_BE( Plen, output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        padlen = 8 * ( ( in_len / 8 ) - 1 ) - Plen;
+        if ( padlen < 0 || padlen > 7 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+        for( i = 0; i < (size_t)padlen; i++ )
+        {
+            if( output[ in_len - padlen + i ] != 0 )
+            {
+                ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+                goto cleanup;
+            }
+        }
+
+        /*
+         * Shift output to point to P
+         */
+        memcpy( output, output + KW_SEMIBLOCK_LENGTH , Plen );
```

Comment:
Please check that `Plen` is no more than `KW_SEMIBLOCK_LENGTH` in order to ensure we don't do a memcpy with overlapping buffers (which is undefined behavior, as the compiler assumes they don't overlap). If you want to allow overlap (that is, `Plen` is still valid if it is more than `KW_SEMIBLOCK_LENGTH`), consider using `memmove()`.